packages/editor/src/optimizer/optimizer.ts

```diff
@@ -0,0 +1,254 @@
+    const requestData: CompileRequestData = {
+      tag: MessageTags.Compile,
+      domain,
+      style,
+      substance,
+      variation,
+    };
+
+    const compileResult = await this.request(requestData);
+    if (isErr(compileResult)) {
+      return compileResult;
+    }
+
+    const diagramId = compileResult.value.diagramId;
+
+    // get the render state, which gets and caches the svg and label measurements for us
+    const renderState = await this.computeLayout(diagramId, {
+      sequenceId: 0,
+      step: 0,
+    });
+    if (renderState.isErr()) {
+      return taggedErr(
+        MessageTags.Compile,
+        isPenroseError(renderState.error)
+          ? renderState.error
+          : runtimeError(showOptimizerError(renderState.error)),
+      );
+    }
+
+    // send the label measurements to the broker with the destination diagramId
+    notifyWorker(this.broker, {
+      tag: MessageTags.LabelMeasurements,
+      diagramId,
+      labelMeasurements: this.labelMeasurements.get(diagramId)!,
+    });
+
+    return compileResult;
+  };
+
+  /**
+   * Get the `HistoryInfo` of the given diagram
+   * @param diagramId
+   */
+  poll = async (diagramId: DiagramID): Promise<PollResult> => {
+    log.info("Sending poll request to broker");
+    const requestData: PollRequestData = {
+      tag: MessageTags.Poll,
+      diagramId,
+    };
+
+    return await this.request(requestData);
+  };
+
+  /**
+   * Compute the `RenderState` of the given diagram and location in history
+   * @param diagramId
+   * @param historyLoc
+   */
+  computeLayout = async (
+    diagramId: DiagramID,
+    historyLoc: HistoryLoc,
+  ): Promise<
+    Result<
+      // actual message returns a `LayoutState`, so we can't return the
+      // message result type directly
+      RenderState,
+      InvalidDiagramIDError | InvalidHistoryLocError | PenroseError
+    >
+  > => {
+    const requestData: ComputeLayoutRequestData = {
+      tag: MessageTags.ComputeLayout,
+      diagramId,
+      historyLoc,
+    };
+    const result = await this.request(requestData);
+    if (isErr(result)) {
+      return Result.err(result.error);
+    }
+
+    let svgCache: Map<string, HTMLElement>;
+
+    // if we haven't already cached the svgs and label measurements, calulate them
+    // and store them
+    if (!this.svgCaches.has(diagramId)) {
+      const svgCacheResult = await collectAndSeparateLabels(
+        result.value.shapes,
+      );
+      if (svgCacheResult.isErr()) {
+        return Result.err(svgCacheResult.error);
+      }
+      svgCache = svgCacheResult.value.svgCache;
+      this.svgCaches.set(diagramId, svgCache);
+      this.labelMeasurements.set(diagramId, svgCacheResult.value.optLabelCache);
+    } else {
+      svgCache = this.svgCaches.get(diagramId)!;
+      this.labelMeasurements.set(diagramId, result.value.labelMeasurements);
+    }
+
+    return Result.ok(layoutStateToRenderState(result.value, svgCache));
+  };
```

Comment:
Funny, something about this function breaks syntax highlighting in VS Code!

Reply:
same here!

Reply:
I'm guessing it's the ambiguity around `>` that caused the faulty highlighting?